util/sym-mode.go

```diff
@@ -0,0 +1,262 @@
+// Mgmt
+// Copyright (C) 2013-2019+ James Shubin and the project contributors
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package util
+
+import (
+	"os"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	ModeUser  uint32 = 64
+	ModeGroup uint32 = 8
+	ModeOther uint32 = 1
+
+	ModeRead  uint32 = 4
+	ModeWrite uint32 = 2
+	ModeExec  uint32 = 1
+
+	ModeSetU   uint32 = 4
+	ModeSetG   uint32 = 2
+	ModeSticky uint32 = 1
+)
+
+func modeIsValidWho(who string) bool {
+	for _, w := range []string{"u", "g", "o"} {
+		who = strings.ReplaceAll(who, w, "")
+	}
+	return len(who) == 0
+}
+
+func modeIsValidWhat(what string) bool {
+	for _, w := range []string{"w", "r", "x", "s", "t"} {
+		what = strings.ReplaceAll(what, w, "")
+	}
+	return len(what) == 0
+}
+
+// SymbolicMode is a function that parses a string of symbolic file mode and
+// applies it to the from FileMode.
+//
+// SymbolicFile mode is expected to be a list of who (user, group, other) then
+// the operation (=, +, -) then the change (read, write, execute, setuid,
+// setgid, sticky).
+//
+// Ex. ug=rw
+func SymbolicMode(mode string, from os.FileMode) (os.FileMode, error) {
+	// parse the who
+	// var i int
+	if strings.Contains(mode, "=") && !strings.ContainsAny(mode, "+-") {
+		return modeAssigned(mode, from)
+	} else if strings.Contains(mode, "+") && !strings.ContainsAny(mode, "=-") {
+		return modeAdded(mode, from)
+	} else if strings.Contains(mode, "-") && !strings.ContainsAny(mode, "=+") {
+		return modeSubtracted(mode, from)
+	}
+
+	return from, errors.Errorf("%s is not a valid symbolic mode", mode)
+}
+
+func modeAssigned(mode string, from os.FileMode) (os.FileMode, error) {
+	var who, what string
+	if sp := strings.Split(mode, "="); len(sp) != 2 {
+		return from, errors.Errorf("invalid mode: expected only one '=' but got %d", len(sp))
+	} else {
+		who, what = sp[0], sp[1]
+	}
+
+	// Validate input
```

Comment:
Short comments are either lower case snippets or full sentences that are capitalized and end with a period. Same comment for other places. https://github.com/purpleidea/mgmt/blob/master/docs/style-guide.md#comments